experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts

```diff
@@ -0,0 +1,144 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { UsageError } from "@fluidframework/telemetry-utils";
+import { fail } from "../util";
+import { EmptyObject } from "../feature-libraries";
+import { SchemaFactory } from "./schemaFactory";
+// This type is referred to by the base type generated by schema.
+// Adding the import of it here prevents TypeScript from generating inline includes for it in the d.ts file.
+// If we let TypeScript generate these includes, they use relative paths which break API extractor's rollup.
+// eslint-disable-next-line unused-imports/no-unused-imports
+import { type NodeKind as _dummy } from "./schemaTypes";
+
+/**
+ * Create a schema for a node with no state.
+ * @remarks
+ * This is commonly used in unions when the only information needed is which kind of node the value is.
+ * Enums are a common example of this pattern.
+ * @alpha
+ */
+export function singletonSchema<TScope extends string, TName extends string>(
+	factory: SchemaFactory<TScope>,
+	name: TName,
+) {
+	return class SingletonSchema extends factory.object(name, {}) {
+		public constructor(data?: EmptyObject) {
+			super(data ?? {});
+		}
+		public get value(): TName {
+			return name;
+		}
+	};
+}
+
+/**
+ * Converts an enum into a collection of schema which can be used in a union.
+ * @alpha
+ */
+export function adaptEnum<TScope extends string, const TEnum extends Record<string, string>>(
+	factory: SchemaFactory<TScope>,
+	members: TEnum,
+) {
+	type Values = TEnum[keyof TEnum];
+	const values = Object.values(members) as Values[];
+	const inverse = new Map(Object.entries(members).map(([key, value]) => [value, key])) as Map<
+		Values,
+		keyof TEnum
+	>;
+
+	if (inverse.size !== values.length) {
+		throw new UsageError("All members of enums must have distinct values.");
+	}
+
+	type TOut = {
+		readonly [Property in keyof TEnum]: ReturnType<
+			typeof singletonSchema<TScope, TEnum[Property]>
+		>;
+	};
+	const factoryOut = <TValue extends Values>(value: TValue) => {
+		return new out[inverse.get(value) ?? fail("missing enum value")](
+			{},
+		) as unknown as ReturnType<typeof singletonSchema<TScope, TValue>>;
+	};
+	const out = factoryOut as typeof factoryOut & TOut;
+	for (const [key, value] of Object.entries(members)) {
+		Object.defineProperty(out, key, {
+			enumerable: true,
+			configurable: false,
+			writable: false,
+			value: singletonSchema(factory, value),
+		});
+	}
+
+	return out;
+}
+
+/**
+ * `Object.values`, but with more specific types.
+ * @alpha
+ */
+export function typedObjectValues<TKey extends string, TValues>(
```

Comment:
Does this need to be exported?

Reply:
Actually, looks like it isn't exported, so we can probably remove the `@alpha` tag

Reply:
The examples need it, so it should be exported as real users will need it as well (ex: to create a field that contains the enum in a schema)

Reply:
Added examples using it to docs.